custom/icds_reports/reports/disha.py

```diff
@@ -34,23 +35,62 @@ def get_data(self):
         else:
             return ""
 
-    def build(self):
+    def _get_columns(self):
         columns = [field.name for field in DishaIndicatorView._meta.fields]
         columns.remove("month")
-        indicators = DishaIndicatorView.objects.filter(
+        return columns
+
+    def _get_rows(self):
+        return DishaIndicatorView.objects.filter(
             month=self.month,
             state_name__iexact=self.state_name
-        ).values_list(*columns)
-        data = {
-            "month": str(self.month),
-            "state_name": self.state_name,
-            "column_names": columns,
-            "rows": list(indicators)
-        }
-        file = BytesIO(json.dumps(data, cls=DjangoJSONEncoder))
-        blob_ref, _ = IcdsFile.objects.get_or_create(blob_id=self._blob_id(), data_type='disha_dumps')
-        blob_ref.store_file_in_blobdb(file, expired=DISHA_DUMP_EXPIRY)
-        blob_ref.save()
+            # batch_qs requires ordered queryset
+        ).order_by('pk').values_list(*self._get_columns())
+
+    def _write_data_in_chunks(self, temp_path):
+        # Writes indicators in json format to the file at temp_path
+        #   in chunks so as to avoid memory errors while doing json.dumps.
+        #   The structure of the json is as below
+        #   {
+        #       'month': '2018-09-01',
+        #       'state_name': 'Andhra Pradesh',
+        #       'columns': [<List of disha columns>],
+        #       'rows': List of lists of rows, in the same order as columns
+        #   }
+        #
+        # The chunk_size should not be too low, otherwise there will be too many
+        #   queries to for each chunk, it should not be too high otherwise
+        #   there would be memory errors while writing json formatted data to file
+        chunk_size = 4000000
+        columns = self._get_columns()
+        indicators = self._get_rows()
+        metadata_line = u'{{'\
+            '"month":"{month}", '\
+            '"state_name": "{state_name}", '\
+            '"column_names": {columns}, '\
+            '"rows": ['.format(
+                month=self.month,
+                state_name=self.state_name,
+                columns=json.dumps(columns, ensure_ascii=False))
+        with open(temp_path, 'w', encoding='utf-8') as f:
+            f.write(metadata_line)
+            written_count = 0
+            for _, end, total, chunk in batch_qs(indicators, chunk_size):
+                chunk_string = json.dumps(list(chunk), ensure_ascii=False)
+                # chunk is list of lists, so skip enclosing brackets
+                f.write(chunk_string[1:-1])
+                written_count += len(chunk)
+                if written_count != total:
+                    f.write(u",")
+            f.write(u"]}")
```

Comment:
So this is very carefully writing a json response to a file in chunks. This looks a bit brittle to me.
I don't have a better solution in mind for the problem but do you intend to look for a better one in coming future?

Reply:
This is brittle, I have spent quite a bit of time exploring a solution. The only other solution are 1. to take a direct db dump, 2. Paginate the endpoint  (but these doesn't address the DISHA requirement) 3. (Another brittle solution) to generate dump in CSV with custom delimiters and concat files.

I had carefully commented out and isolated code for tests for this reason.

Reply:
I myself don't have suggestion on what a better approach can be here, but a bigger group might have ways to do this in a more secure manner. So this is fine for now but meanwhile might be worth to check with others for suggestions.

Reply:
👍